File: sixthSeminar/src/main/java/org/sopt/sixthSeminar/service/PostService.java
package org.sopt.sixthSeminar.service;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.sopt.sixthSeminar.domain.Category;
import org.sopt.sixthSeminar.domain.CategoryId;
import org.sopt.sixthSeminar.domain.Member;
import org.sopt.sixthSeminar.domain.Post;
import org.sopt.sixthSeminar.domain.ServiceMember;
import org.sopt.sixthSeminar.dto.request.post.PostCreateRequest;
import org.sopt.sixthSeminar.dto.request.post.PostUpdateRequest;
import org.sopt.sixthSeminar.dto.response.post.PostGetResponse;
import org.sopt.sixthSeminar.exception.ErrorMessage;
import org.sopt.sixthSeminar.exception.model.NotFoundException;
import org.sopt.sixthSeminar.repository.PostRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
@Slf4j
@Transactional(readOnly = true)
public class PostService {

    private final PostRepository postRepository;
    private final ServiceMemberService serviceMemberService;
    private final CategoryService categoryService;


    @Transactional
    public void editContent(Long postId, PostUpdateRequest request) {
        Post post = postRepository.findById(postId)
                .orElseThrow(() -> new NotFoundException(ErrorMessage.POST_NOT_FOUND_EXCEPTION));
        post.updateContent(request.content());
    }

    @Transactional
    public void deleteById(Long postId) {
        Post post = postRepository.findById(postId).orElseThrow(() -> new NotFoundException(ErrorMessage.POST_NOT_FOUND_EXCEPTION));
        postRepository.delete(post);
    }

    public List<PostGetResponse> getPosts(Long memberId) {
        return postRepository.findAllByServiceMember(serviceMemberService.findById(memberId))
                .stream()
                .map(post -> PostGetResponse.of(post, getCategoryByPost(post)))
                .toList();
    }

    public PostGetResponse getById(Long postId) {
        Post post = postRepository.findById(postId).orElseThrow(() -> new NotFoundException(ErrorMessage.POST_NOT_FOUND_EXCEPTION));
        return PostGetResponse.of(post, getCategoryByPost(post));
    }

    private Category getCategoryByPost(Post post) {
        return categoryService.getByCategoryId(post.getCategoryId());
    }

}

File: sixthKotlin/src/main/kotlin/org/sopt/sixthKotlin/service/PostService.kt
package org.sopt.sixthKotlin.service

import org.sopt.sixthKotlin.controller.dto.request.PostCreateRequest
import org.sopt.sixthKotlin.domain.Post
import org.sopt.sixthKotlin.exception.ErrorMessage
import org.sopt.sixthKotlin.exception.SoptException.PostNotFoundException
import org.sopt.sixthKotlin.external.S3Service
import org.sopt.sixthKotlin.repository.PostJpaRepository
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.multipart.MultipartFile
import java.io.IOException

@Service
@Transactional(readOnly = true)
class PostService(
    private val serviceMemberService: ServiceMemberService,
    private val postJpaRepository: PostJpaRepository,
    private val s3Service: S3Service
) {
    @Transactional
    fun create(
        request: PostCreateRequest,
        image: MultipartFile?,
        memberId: Long?
    ): String {
        return try {
            val imageUrl = s3Service.uploadImage(POST_IMAGE_FOLDER_NAME, image!!)
            val serviceMember = serviceMemberService.findById(memberId!!)
            val post = postJpaRepository.save(
                Post(
                    request.title,
                    request.content,
                    imageUrl,
                    serviceMember
                )
            )
            post.id.toString()
        } catch (e: RuntimeException) {
            throw RuntimeException(e.message)
        } catch (e: IOException) {
            throw RuntimeException(e.message)
        }
    }

    @Transactional
    fun deleteById(postId: Long) {
        try {
            val post =
                postJpaRepository.findByIdOrNull(postId) ?: throw PostNotFoundException(ErrorMessage.POST_NOT_FOUND)
            s3Service.deleteImage(post.imageUrl)
            postJpaRepository.deleteById(postId)
        } catch (e: IOException) {
            throw RuntimeException(e.message)
        } catch (e: RuntimeException) {
            throw RuntimeException(e.message)
        }
    }

    companion object {
        private const val POST_IMAGE_FOLDER_NAME = "posts/"
    }
}
